extensions/amp-story-shopping/0.1/amp-story-shopping-config.js

```diff
@@ -1,26 +1,55 @@
-import {Layout, applyFillContent} from '#core/dom/layout';
+import {Layout} from '#core/dom/layout';
 
-const TAG = 'amp-story-shopping-config';
+import {Services} from '#service';
+
+import {devAssert} from '#utils/log';
+
+import {Action} from '../../amp-story/1.0/amp-story-store-service';
 
 export class AmpStoryShoppingConfig extends AMP.BaseElement {
   /** @param {!AmpElement} element */
   constructor(element) {
     super(element);
+    /** @private @const {?./amp-story-store-service.AmpStoryStoreService} */
+    this.storeService_ = null;
+
+    /** @private @const {!./amp-story-request-service.AmpStoryRequestService} */
+    this.requestService_ = null;
+  }
+
+  /**
+   * Keys product data to product-ids and adds them to the store service.
+   * @param {!JsonObject} storyConfig
+   * @private
+   */
+  addShoppingStateFromConfig_(storyConfig) {
+    const productIDtoProduct = {};
+
+    for (const item of storyConfig['items']) {
+      productIDtoProduct[item['product-tag-id']] = item;
+    }
 
-    /** @private {string} */
-    this.myText_ = TAG;
+    this.storeService_.dispatch(Action.ADD_SHOPPING_DATA, productIDtoProduct);
 
-    /** @private {?Element} */
-    this.container_ = null;
+    //TODO(#36412): Add call to validate config here.
   }
 
   /** @override */
   buildCallback() {
     super.buildCallback();
-    this.container_ = this.element.ownerDocument.createElement('div');
-    this.container_.textContent = this.myText_;
-    this.element.appendChild(this.container_);
-    applyFillContent(this.container_, /* replacedContent */ true);
+    return Promise.all([
+      Services.storyStoreServiceForOrNull(this.win).then((storeService) => {
+        devAssert(storeService, 'Could not retrieve AmpStoryStoreService');
+        this.storeService_ = storeService;
+      }),
+      Services.storyRequestServiceForOrNull(this.win).then((service) => {
+        this.requestService_ = service;
+      }),
+    ])
+      .then(() => {
+        return this.requestService_.loadConfigImpl(this.element);
+      })
+      .then((storyConfig) => this.addShoppingStateFromConfig_(storyConfig));
```

Comment:
We need to return a promise so that buildCallback resolves.
Not returning a promise in`buildCallback` could cause race conditions since it will block the lifecycle methods of the component.

Reply:
We are returning a promise in buildCallback right? 
I'm not sure how to get this to work other than the current setup. 
Can you please clarify how you would structure this code with a working example.

Reply:
For sure! We just need to trigger the resolve :)
In this block it could be:
```
.then((storyConfig) => {
  this.addShoppingStateFromConfig_(storyConfig))
  return Promise.resolve()
};
```

The other way would be in `addShoppingStateFromConfig_` by returning Promise.resolve() there and leaving this block unchanged. Either way works :)

Reply:
I think we actually don't need to return `Promise.resolve()` here. When nothing is returned within `.then()`, the promise automatically gets resolved with an `undefined` value 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then#return_value

Reply:
Sounds good to me - I have removed the promise.resolve unless there are any objections.